webapp/graphite/storage.py

```diff
@@ -331,6 +371,127 @@ def distance_to_requested_interval(node):
             reader = MultiReader(minimal_node_set)
             return LeafNode(path, reader)
 
+    def tagdb_auto_complete_tags(self, exprs, tagPrefix=None, limit=None, requestContext=None):
+        log.debug(
+            'graphite.storage.Store.auto_complete_tags :: Starting lookup on all backends')
+
+        if requestContext is None:
+          requestContext = {}
+
+        jobs = []
+        use_tagdb = False
+        for finder in self.get_finders(requestContext.get('localOnly')):
+          if getattr(finder, 'tags', False):
+            jobs.append(Job(finder.auto_complete_tags, exprs, tagPrefix=tagPrefix, limit=limit, requestContext=requestContext))
+          else:
+            use_tagdb = True
+
+        if not jobs:
+          if not use_tagdb:
+            return []
+
+          return self.tagdb.auto_complete_tags(exprs, tagPrefix=tagPrefix, limit=limit, requestContext=requestContext)
+
+        # start finder jobs
+        jobs = self.pool_exec(jobs, settings.REMOTE_FIND_TIMEOUT)
+
+        results = set()
+
+        # if we're using the local tagdb then execute it
+        if use_tagdb:
+          results.update(self.tagdb.auto_complete_tags(exprs, tagPrefix=tagPrefix, limit=limit, requestContext=requestContext))
+
+        done = 0
+        errors = 0
+
+        # Start fetches
+        start = time.time()
+        try:
+          for job in jobs:
+            done += 1
+
+            if job.exception:
+              errors += 1
+              raise Exception('%s %s %s %s' % (job.func, job.args, job.kwargs, job.exception))
+              log.info("Autocomplete tags for %s %s failed after %fs: %s" % (str(exprs), tagPrefix or '', time.time() - start, str(job.exception)))
```

Comment:
should it be before raise Exception?

Reply:
That is some testing code that got through, will clean it up when I get the auto-complete tests fleshed out.